### PR TITLE
Use Golang

### PR DESCRIPTION
**Pull request by jamierocks**
- [x] explanation on README
- [x] pull announcements (#15) - http://i.imgur.com/UUyJ5Fp.png
- [x] make port configurable
- [x] make announcements look nicer - http://i.imgur.com/y7lbyBg.png

### CODE REVIEW

**Comment by progwml6** [comment]
I think Sass partials would be a good idea as well

how close are these templates to what we have in docs or ore?

**Comment by jamierocks** [comment]
Well I was going to see if people want to stick with SASS or switch to scss/less/whatever first :P 

Ore uses Scala, so it's rather different, however it should be similar to the docs.

**Comment by Tzky** [comment]
Docs use Stylus and Jinja. Example files:
Jinja: https://github.com/SpongePowered/sponge_docs_theme/blob/master/src/html/layout.html
Stylus: https://github.com/SpongePowered/sponge_docs_theme/blob/master/src/styl/homepage.styl

No clue about Ore as i haven't had a proper look at the code yet.

**Comment by progwml6** [comment]
Is it plausible to switch docs to a different css flavor if we wanted to?

**Comment by jamierocks** [comment]
Yes.

File: controllers/announcements.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    Link string `json:"link"`
+}
+
+func GetAnnouncements(ctx *macaron.Context) {
+    data := url.Values{}
+    client := &http.Client{}
+    r, _ := http.NewRequest("GET",
+        "https://forums.spongepowered.org/c/announcements.json?order=created", bytes.NewBufferString(data.Encode()))
+
+    resp, _ := client.Do(r)
```
Comment (stephan-gh) [inline controllers/announcements.go]: You should likely do something with the error, it's not a good idea to simply eat it.

Comment (jamierocks) [inline controllers/announcements.go]: Yeah I should :P 

i was doing this rather quick to get a proof of concept :P

File: controllers/base.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package controllers
+
+import (
+    "gopkg.in/macaron.v1"
+)
+
+func GetHomepage(ctx *macaron.Context) {
+    ctx.HTML(200, "index")
```
Comment (stephan-gh) [inline controllers/base.go]: It's always good to use pre-defined constants, there is `http.StatusOK`

File: controllers/announcements.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+type Announcement struct {
+    Title string `json:"title"`
+    Content string `json:"content"`
+    Link string `json:"link"`
+}
+
+func GetAnnouncements(ctx *macaron.Context) {
+    data := url.Values{}
+    client := &http.Client{}
```
Comment (stephan-gh) [inline controllers/announcements.go]: Why do you need a custom client if you don't set anything (from what I can tell), the `http` package has a default client you can use, together with a useful global method `http.Get(url)`

Comment (jamierocks) [inline controllers/announcements.go]: Awesome. You learn something new everyday :P

File: controllers/base.go (hunk 2 above)
Comment (jamierocks) [inline controllers/base.go]: I'll get to that.

File: controllers/announcements.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        Second: second,
+    })
+}
+
+func getAnnouncement(title string, id int, slug string) (Announcement, error) {
+    r, err := http.Get("https://forums.spongepowered.org/t/" + strconv.Itoa(id) + ".json")
+    body, err := ioutil.ReadAll(r.Body)
+
+    if err != nil {
+        return *&Announcement{}, err
```
Comment (stephan-gh) [inline controllers/announcements.go]: Lol :P You create an `Announcement`, take a pointer of it, and dereference it directly again, you can just do `Annoucement{}` directly then.

**Comment by Tzky** [comment]
yep, docs theme is build with gulp and node.js too, so changing it isn't a big task.

File: controllers/announcements.go (hunk 4 above)
Comment (jamierocks) [inline controllers/announcements.go]: Yeah not sure what I was doing there XD

File: controllers/announcements.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+type Announcement struct {
+    Title string `json:"title"`
+    Content string `json:"content"`
+    Link string `json:"link"`
+}
+
+func GetAnnouncements(ctx *macaron.Context) {
+    r, err := http.Get("https://forums.spongepowered.org/c/announcements.json?order=created")
+    body, err := ioutil.ReadAll(r.Body)
```
Comment (stephan-gh) [inline controllers/announcements.go]: No need to read + buffer everything first, there is a `json.Decoder` you can use directly on the body, see [http://blog.golang.org/json-and-go](http://blog.golang.org/json-and-go#TOC_7.)

**Comment by jamierocks** [comment]
How do you feel about switching to SCSS @Tzky ?

File: controllers/announcements.go (hunk 5 above)
Comment (jamierocks) [inline controllers/announcements.go]: Oh awesome.

File: controllers/announcements.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return
+    }
+
+    ctx.JSON(http.StatusOK, &AnnouncementView{
+        First: *first,
+        Second: *second,
+    })
+}
+
+func getAnnouncement(title string, id int, slug string) (*Announcement, error) {
```
Comment (stephan-gh) [inline controllers/announcements.go]: You can save even more code by making this take a `Topic` directly

File: controllers/announcements.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    Title string `json:"title"`
+    Content string `json:"content"`
+    Link string `json:"link"`
+}
+
+func GetAnnouncements(ctx *macaron.Context) {
+    var res Category
+
+    r, err := http.Get("https://forums.spongepowered.org/c/announcements.json?order=created")
+    err = json.NewDecoder(r.Body).Decode(&res)
```
Comment (stephan-gh) [inline controllers/announcements.go]: Need to check for error before this line.

Comment (jamierocks) [inline controllers/announcements.go]: actually yes - not sure why I did that :P

File: templates/base.html [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <meta name="description" content="">
+    <meta name="author" content="">
+
+    <title>Sponge - Minecraft Modding API</title>
+    <link rel="icon" type="image/png" href="https://forums-cdn.spongepowered.org/uploads/default/original/2X/9/9ba706a80e45cf427617525ee2a19fad7bb6b109.png">
+
+    <!-- Bootstrap Core CSS - Uses Bootswatch Flatly Theme: https://bootswatch.com/flatly/ -->
+    <link href="/assets/css/bootstrap.min.css" rel="stylesheet">
```
Comment (kashike) [inline templates/base.html:15]: Can this not be a CDN href? Are there local changes?

File: templates/base.html [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    <!-- Custom CSS -->
+    <link href="/assets/css/spongehome.min.css" rel="stylesheet">
+
+    <!-- Custom Fonts -->
+    <link href="/assets/font-awesome/css/font-awesome.min.css" rel="stylesheet" type="text/css">
```
Comment (kashike) [inline templates/base.html:21]: Can this not be a CDN href? Are there local changes?

File: templates/base.html [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                <div class="col-lg-12">
+                    Copyright &copy; SpongePowered 2014-2016
+                </div>
+            </div>
+        </div>
+    </div>
+</footer>
+
+<!-- jQuery -->
+<script src="/assets/js/jquery.js"></script>
```
Comment (kashike) [inline templates/base.html:113]: Can this not be a CDN href? Are there local changes?

File: templates/base.html [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+<!-- Bootstrap Core JavaScript -->
+<script src="/assets/js/bootstrap.min.js"></script>
```
Comment (kashike) [inline templates/base.html:116]: Can this not be a CDN href? Are there local changes?

File: templates/base.html (hunk 10 above)
Comment (jamierocks) [inline templates/base.html:113]: @Owexz could I use cdnjs for everything?

File: controllers/announcements.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        return
+    }
+
+    ctx.JSON(http.StatusOK, &AnnouncementView{
+        First: *first,
+        Second: *second,
+    })
+}
+
+func getAnnouncement(topic Topic) (*Announcement, error) {
```
Comment (stephan-gh) [inline controllers/announcements.go]: Do you really need a pointer here?

Comment (jamierocks) [inline controllers/announcements.go]: well you moaned about the opposite :p I do prefer this way though

Comment (stephan-gh) [inline controllers/announcements.go]: I "moaned" about taking the pointer and dereferencing it again directly after, you can just do `return Annoucement{}, nil`

**Comment by Tzky** [comment]
I like scss and Stylus. Stylus looks more beautiful but is less powerful than scss afaik

File: controllers/base.go [hunk 13]
```diff
@@ -0,0 +1,74 @@
+package controllers
+
+import (
+    "net/http"
+    "gopkg.in/macaron.v1"
+)
+
+var (
+    SPONSORS []Sponsor = []Sponsor{
```
Comment (stephan-gh) [inline controllers/base.go]: Correct me if I'm wrong but I think Go's convention is `Sponsors`, not full caps

Comment (jamierocks) [inline controllers/base.go]: It is - that'll probably be my cap lock that doesn't always work :|

**Comment by jamierocks** [comment]
https://github.com/SpongePowered/SpongeHome/commit/6a237724e7e782b6220811473f6708ee55bd301c

**Pull request closed without merging**